cpp/src/arrow/dataset/scanner.cc

```diff
@@ -58,6 +55,62 @@ static ScanTaskIterator GetScanTaskIterator(DataFragmentIterator fragments,
   return MakeFlattenIterator(std::move(maybe_scantask_it));
 }
 
+template <typename Fn>
+class MapScanTask : public ScanTask {
+ public:
+  MapScanTask(std::unique_ptr<ScanTask> task, Fn map)
+      : task_(std::move(task)), map_(std::move(map)) {}
+
+  RecordBatchIterator Scan() override {
+    return MakeMaybeMapIterator(map_, task_->Scan());
+  }
+
+ private:
+  std::unique_ptr<ScanTask> task_;
+  Fn map_;
+};
+
+static ScanTaskIterator FilterScanTask(ScanTaskIterator it,
+                                       std::shared_ptr<ExpressionEvaluator> evaluator,
+                                       std::shared_ptr<Expression> filter) {
+  if (evaluator == nullptr || filter == nullptr) {
```

Comment:
Neither filter nor evaluator should be nullptr; their default values are scalar(true) and NoopEvauator() respectively, so this null check can be removed

Reply:
To check for a trivial filter expression, use
```suggestion
  if (filter->Equals(true)) {
```